### PR TITLE
Added bubble, stacked bar chart components and services

### LINKED ISSUE #395: Adding visualisations with angular directives & d3
Using directives for each chart will help to organise different parts of front-end apps into components, and make them reusable in other apps. I aim to create a bubble chart and stacked bar chart directive for milestone 2.

- [x]  Bubble Chart
- [x]  Stacked Bar Chart

### LINKED ISSUE #411: Parsing json response by date for visualisations
Related to #391, to analyse by date in the visualisations the json response has to be processed by date.

### PR DESCRIPTION
**Pull request by leonmak**
Fixes #395, created 2 bubble charts for mentions and word count using the bubble chart directive and some services for parsing the tweet array, and a stacked bar chart for top hashtags/day using a service to group each tweet by date, fixes #411 .

### CODE REVIEW

File: html/apps/bubblecharts/charts/bubbleCloud.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+.directive('bubbleCloud', [ 
+
+  function () {
+
+    var link = function ($scope, $el, $attrs) {
+        
+        var diameter = 500;
+
+        var id = '#bubbleCloud_'+$scope.$id;
```
Comment (sudheesh001) [inline html/apps/bubblecharts/charts/bubbleCloud.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2618402745/issues/source?bid=3288751&fileBranchId=3313532#l25)

File: html/apps/bubblecharts/charts/bubbleCloud.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                return;
+            }
+            // if ($scope.mentionFreq) chart.select("#loading").remove();
+            
+            
+            function processData(obj) {
+                var newDataSet = [];
+                for(var prop in obj) {
+                    if(obj[prop] >= $scope.min){
+                        var node = {name: prop, className: 'bubbles', size: obj[prop]}
```
Comment (sudheesh001) [inline html/apps/bubblecharts/charts/bubbleCloud.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2618402745/issues/source?bid=3288751&fileBranchId=3313532#l77)

File: html/apps/bubblecharts/charts/bubbleCloud.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
+(function(){
+angular.module('myApp')
```
Comment (sudheesh001) [inline html/apps/bubblecharts/charts/bubbleCloud.js]: Codacy Issue found: ["angular" is not defined.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2618402745/issues/source?bid=3288751&fileBranchId=3313532#l2)

File: html/apps/bubblecharts/charts/bubbleCloud.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                .transition().duration(1000)
+                .attr("opacity", 0)
+                
+                chart.selectAll("circle")
+                .transition().duration(1000)
+                .attr("r", 0)
+                
+                window.setTimeout(function(){
+                    // select all bubble groups, remove data then remove dom nodes
+                    chart.selectAll('.bubble').data([])
```
Comment (sudheesh001) [inline html/apps/bubblecharts/charts/bubbleCloud.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2618402745/issues/source?bid=3288751&fileBranchId=3313532#l64)

File: html/apps/bubblecharts/charts/bubbleCloud.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
+(function(){
+angular.module('myApp')
+
+.controller('bubbleCloudCtrl', ['$scope','$window','$localStorage',
```
Comment (sudheesh001) [inline html/apps/bubblecharts/charts/bubbleCloud.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2618402745/issues/source?bid=3288751&fileBranchId=3313532#l4)

File: html/apps/bubblecharts/charts/bubbleCloud.js [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        var bubble = d3.layout.pack()
+            .size([diameter, diameter])
+            .value(function(d) {return d.size;}) // more mentions, bigger bubble
+            .sort(function(a, b) { // biggest in the middle
+                return -(a.value - b.value)
+            }) 
+            .padding(3);
+        
+        // remember to set to true to watch values, if not $watch will not run as reference same
+        $scope.$watch('data', update, true);
```
Comment (sudheesh001) [inline html/apps/bubblecharts/charts/bubbleCloud.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2618402745/issues/source?bid=3288751&fileBranchId=3313532#l46)

File: html/apps/bubblecharts/charts/bubbleCloud.js [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        
+        function update(mentionFreq) {
+            console.log("updating");
+            console.log(mentionFreq);
+            // if new poll, don't enter, fade out everything
+            if(!mentionFreq || Object.keys(mentionFreq).length === 0) {
+                
+                chart.selectAll("text")
+                .transition().duration(1000)
+                .attr("opacity", 0)
```
Comment (sudheesh001) [inline html/apps/bubblecharts/charts/bubbleCloud.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2618402745/issues/source?bid=3288751&fileBranchId=3313532#l56)

File: html/apps/bubblecharts/charts/bubbleCloud.js [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            // generate data with calculated layout values
+            var nodes = bubble.nodes(processData(mentionFreq))
+                .filter(function(d) { return !d.children; }); 
+            // filter out the outer bubble for pack layout
+            console.log(nodes);
+            
+            var vis = chart.selectAll('.bubble')
+            .data(nodes);
+            
+            console.log(vis);
```
Comment (sudheesh001) [inline html/apps/bubblecharts/charts/bubbleCloud.js]: Codacy Issue found: [Unexpected console statement.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2618402745/issues/source?bid=3288751&fileBranchId=3313532#l93)

File: html/apps/bubblecharts/charts/bubbleCloud.js [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+            
+            // add attr to each entered group consisting of <circle> and <text>
+            var enter = vis.enter().append('g')
+            .attr('class', 'bubble')
+            
+            enter.append('circle')
```
Comment (sudheesh001) [inline html/apps/bubblecharts/charts/bubbleCloud.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2618402745/issues/source?bid=3288751&fileBranchId=3313532#l99)

File: html/apps/bubblecharts/charts/bubbleCloud.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+  function () {
+
+    var link = function ($scope, $el, $attrs) {
+        
+        var diameter = 500;
+
+        var id = '#bubbleCloud_'+$scope.$id;
+
+        
+        var svg = d3.select($el[0]).append('svg')
```
Comment (sudheesh001) [inline html/apps/bubblecharts/charts/bubbleCloud.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2618402745/issues/source?bid=3288751&fileBranchId=3313532#l28)

**Comment by Orbiter** [comment]
@sudheesh001 please turn that "Strings must use doublequote" off - http://www.w3schools.com/js/js_strings.asp explains that single- and double-Quote is valid code.

**Comment by sudheesh001** [comment]
@Orbiter That was a default rule on codacy, disabled it now 😄

File: html/apps/barchart/services/analyticService.js [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+                    counts[sWord] = counts[sWord] || 0;
+                    counts[sWord]++;
+                }
+            }
+            // // console.log("counts: ", counts);
+            
+            if(Object.keys(counts).length === 0 && counts.constructor === Object){
+                $localStorage.wordFreq = counts;
+            } else {
+                for (sWord in counts){
```
Comment (sudheesh001) [inline html/apps/barchart/services/analyticService.js]: Codacy Issue found: ["sWord" used outside of binding context.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2626095151/issues/source?bid=3288751&fileBranchId=3313532#l81)

File: html/apps/barchart/services/analyticService.js [hunk 12]
```diff
@@ -0,0 +1,202 @@
+angular.module('myApp')
+.factory('analyticService', 
+    function($localStorage, $moment){
+        // returns an object of {word:freq} pairs
+        
+        return {
+            getNewTweets: getNewTweets,
+            updateWordFreq: updateWordFreq,
+            updateMentionFreq: updateMentionFreq,
+            updateHashtagDateFreq: updateHashtagDateFreq
```
Comment (sudheesh001) [inline html/apps/barchart/services/analyticService.js]: Codacy Issue found: ["updateHashtagDateFreq" was used before it was defined](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2626095151/issues/source?bid=3288751&fileBranchId=3313532#l10)

File: html/apps/barchart/services/analyticService.js [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+                    counts[sWord]++;
+                }
+            }
+            // // console.log("counts: ", counts);
+            
+            if(Object.keys(counts).length === 0 && counts.constructor === Object){
+                $localStorage.wordFreq = counts;
+            } else {
+                for (sWord in counts){
+                    $localStorage.wordFreq[sWord] = $localStorage.wordFreq[sWord] || 0;
```
Comment (sudheesh001) [inline html/apps/barchart/services/analyticService.js]: Codacy Issue found: ["sWord" used outside of binding context.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2626095151/issues/source?bid=3288751&fileBranchId=3313532#l82)

File: html/apps/barchart/services/analyticService.js [hunk 14]
```diff
@@ -0,0 +1,202 @@
+angular.module('myApp')
+.factory('analyticService', 
+    function($localStorage, $moment){
+        // returns an object of {word:freq} pairs
+        
+        return {
+            getNewTweets: getNewTweets,
+            updateWordFreq: updateWordFreq,
```
Comment (sudheesh001) [inline html/apps/barchart/services/analyticService.js]: Codacy Issue found: ["updateWordFreq" was used before it was defined](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2626095151/issues/source?bid=3288751&fileBranchId=3313532#l8)

Comment (sudheesh001) [inline html/apps/barchart/services/analyticService.js]: Codacy Issue found: [Expected property shorthand.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2626095151/issues/source?bid=3288751&fileBranchId=3313532#l8)

File: html/apps/barchart/services/analyticService.js [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+            updateMentionFreq: updateMentionFreq,
```
Comment (sudheesh001) [inline html/apps/barchart/services/analyticService.js]: Codacy Issue found: [Expected property shorthand.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2626095151/issues/source?bid=3288751&fileBranchId=3313532#l9)

File: html/apps/barchart/services/analyticService.js [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+            })
+            // set latest day at idx 0 (as dayDiff = 0)
+            .sortKeys(function(a,b){ return b-a;}) 
+            // process tweets each dayDiff
+            .rollup(function(leaves){ return getHashtagFreq(leaves); })
+            .entries(newTweets);
+            // console.log("grouped", groupByDate);
+            
+            // sort desc freq
+            var filteredFreq = d3.nest()
```
Comment (sudheesh001) [inline html/apps/barchart/services/analyticService.js]: Codacy Issue found: ["d3" is not defined.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2626095151/issues/source?bid=3288751&fileBranchId=3313532#l192)

File: html/apps/barchart/services/analyticService.js [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+                var mentionsArr = newTweets.map(function(tweet){
+                    return tweet.mentions;
+                }).reduce(function(prev, next){
+                    return prev.concat(next);
+                }, []);
+                
+                mentionsArr.forEach(function(mention){
+                    $localStorage.mentionFreq[mention] = $localStorage.mentionFreq[mention] || 0;
+                    $localStorage.mentionFreq[mention] ++;
+                })
```
Comment (sudheesh001) [inline html/apps/barchart/services/analyticService.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2626095151/issues/source?bid=3288751&fileBranchId=3313532#l116)

File: html/apps/barchart/services/analyticService.js [hunk 18]
```diff
@@ -0,0 +1,202 @@
+angular.module('myApp')
```
Comment (sudheesh001) [inline html/apps/barchart/services/analyticService.js]: Codacy Issue found: ["angular" is not defined.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2626095151/issues/source?bid=3288751&fileBranchId=3313532#l1)

File: html/apps/barchart/services/analyticService.js [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+.factory('analyticService', 
+    function($localStorage, $moment){
+        // returns an object of {word:freq} pairs
+        
+        return {
+            getNewTweets: getNewTweets,
+            updateWordFreq: updateWordFreq,
+            updateMentionFreq: updateMentionFreq,
+            updateHashtagDateFreq: updateHashtagDateFreq
+        }
```
Comment (sudheesh001) [inline html/apps/barchart/services/analyticService.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2626095151/issues/source?bid=3288751&fileBranchId=3313532#l11)

**Pull request merged**